File: api/routers/jobs.py
"""ワークスペースジョブ / 共通ジョブの CRUD エンドポイント。

ジョブ実行（/run）は `job_runner.py`、共通ロジックは `jobs_common.py` を参照。
"""

import logging

from fastapi import APIRouter, Depends

from ..auth import verify_token
from ..common import GLOBAL_CONFIG_KEY, resolve_workspace_path
from ..config import iter_workspace_entries_from, load_all_config
from ..errors import not_found
from .jobs_common import (
    JobRequest,
    ReorderJobsRequest,
    common_jobs_context,
    delete_job,
    entry_to_job_definition,
    get_workspace_jobs,
    job_definition_to_dict,
    load_common_jobs_data,
    parse_jobs_data,
    reorder_jobs,
    save_job,
    serialize_workspace_jobs,
    ws_jobs_context,
)
from .workspaces import _dynamic_worktree_entries

logger = logging.getLogger(__name__)

router = APIRouter(dependencies=[Depends(verify_token)])


def _merge_jobs_for_ws(display_name, ws_jobs_data, common_jobs_data):
    merged = {}
    for is_common, jobs_data in [(True, common_jobs_data), (False, ws_jobs_data)]:
        for jname, jentry in jobs_data.items():
            merged[jname] = (entry_to_job_definition(jname, jentry), is_common)
    return display_name, {
        jname: job_definition_to_dict(jdef, is_common=is_common)
        for jname, (jdef, is_common) in merged.items()
    }


@router.get("/jobs/workspaces")
def list_all_workspace_jobs():
    all_config = load_all_config()
    common_jobs_data = all_config.get(GLOBAL_CONFIG_KEY, {}).get("jobs", {})
    entries = list(iter_workspace_entries_from(all_config))
    jobs_by_name = {(entry.get("name") or ws_id): entry.get("jobs", {}) for ws_id, entry in entries}
    result = {}
    for ws_id, entry in sorted(entries):
        base = entry.get("worktree_base")
        ws_jobs_data = jobs_by_name[base] if base and base in jobs_by_name else entry.get("jobs", {})
        display_name, serialized = _merge_jobs_for_ws(entry.get("name") or ws_id, ws_jobs_data, common_jobs_data)
        result[display_name] = serialized
    for wt in _dynamic_worktree_entries():
        if wt["name"] in result:
            continue
        wt_jobs = jobs_by_name.get(wt.get("worktree_base", ""), {})
        _, serialized = _merge_jobs_for_ws(wt["name"], wt_jobs, common_jobs_data)
        result[wt["name"]] = serialized
    return result


@router.get("/workspaces/{name}/jobs")
def list_workspace_jobs(name: str):
    resolve_workspace_path(name)
    return serialize_workspace_jobs(name)


@router.get("/workspaces/{name}/jobs/{job_name}")
def get_workspace_job(name: str, job_name: str):
    resolve_workspace_path(name)
    jobs = get_workspace_jobs(name)
    entry = jobs.get(job_name)
    if not entry:
        raise not_found(f"Job '{job_name}' not found")
    job_def, _ = entry
    return job_definition_to_dict(job_def)


@router.post("/workspaces/{name}/jobs")
def create_workspace_job(name: str, body: JobRequest):
    data, save_fn, _ = ws_jobs_context(name)
    return save_job(data, save_fn, None, body, "job created workspace=%s job=%%s" % name)


@router.put("/workspaces/{name}/job-order")
def reorder_workspace_jobs(name: str, body: ReorderJobsRequest):
    data, save_fn, _ = ws_jobs_context(name)
    return reorder_jobs(data, save_fn, body.order, "jobs reordered workspace=%s count=%%d" % name)


@router.put("/workspaces/{name}/jobs/{job_name}")
def update_workspace_job(name: str, job_name: str, body: JobRequest):
    data, save_fn, label = ws_jobs_context(name)
    if job_name not in data:
        raise not_found(f"{label} '{job_name}' not found")
    return save_job(data, save_fn, job_name, body, "job updated workspace=%s job=%%s" % name)


@router.delete("/workspaces/{name}/jobs/{job_name}")
def delete_workspace_job(name: str, job_name: str):
    data, save_fn, label = ws_jobs_context(name)
    return delete_job(data, save_fn, job_name,
                      f"{label} '{job_name}' not found",
                      "job deleted workspace=%s job=%%s" % name)


@router.get("/common/jobs")
def list_common_jobs():
    data = load_common_jobs_data()
    jobs = parse_jobs_data(data)
    return {jname: job_definition_to_dict(jdef) for jname, jdef in jobs.items()}


@router.post("/common/jobs")
def create_common_job(body: JobRequest):
    data, save_fn, _ = common_jobs_context()
    return save_job(data, save_fn, None, body, "common job created job=%s")


@router.put("/common/jobs/{job_name}")
def update_common_job(job_name: str, body: JobRequest):
    data, save_fn, label = common_jobs_context()
    if job_name not in data:
        raise not_found(f"{label} '{job_name}' not found")
    return save_job(data, save_fn, job_name, body, "common job updated job=%s")


@router.delete("/common/jobs/{job_name}")
def delete_common_job(job_name: str):
    data, save_fn, label = common_jobs_context()
    return delete_job(data, save_fn, job_name,
                      f"{label} '{job_name}' not found",
                      "common job deleted job=%s")


@router.put("/common/job-order")
def reorder_common_jobs(body: ReorderJobsRequest):
    data, save_fn, _ = common_jobs_context()
    return reorder_jobs(data, save_fn, body.order, "common jobs reordered count=%d")

File: server/src/jobs.rs
//! ワークスペースジョブ / 共通ジョブの CRUD（Python 側 `api/routers/jobs.py` の移植）。
//!
//! ジョブ実行は `/run`（`job_runner.rs`）と `/dispatch`（`dispatch.rs`）が担う。
//! ここは config.json のジョブ定義 CRUD と一覧のみ。

use std::sync::Arc;

use axum::extract::{Path, State};
use axum::Json;
use serde_json::{Map, Value};

use crate::auth::RequireAuth;
use crate::config::GLOBAL_CONFIG_KEY;
use crate::errors::{not_found, ApiError};
use crate::git_utils::{dynamic_worktree_entries, resolve_workspace_path};
use crate::jobs_common::{
    commit_common_jobs, commit_workspace_jobs, delete_job, load_common_jobs_data,
    load_workspace_jobs_data, merge_jobs_serialized, reorder_jobs, resolve_jobs_owner, save_job,
    serialize_workspace_jobs, JobRequest, ReorderJobsRequest,
};
use crate::state::AppState;
use crate::util::JsonBody;

// ─── GET /jobs/workspaces ───────────────────────────────────────────────────

pub async fn list_all_workspace_jobs(
    State(state): State<Arc<AppState>>,
    _auth: RequireAuth,
) -> Result<Json<Value>, ApiError> {
    let all_config = state.config.load_all();
    let common_jobs = all_config
        .get(GLOBAL_CONFIG_KEY)
        .and_then(|g| g.get("jobs"))
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();
    // 1回の走査で表示名・worktree_base・自前ジョブを集める。表示名 -> jobs は
    // worktree_base 解決用（登録済み・動的 worktree の両方が参照する）。
    let mut jobs_by_name: std::collections::HashMap<String, Map<String, Value>> =
        std::collections::HashMap::new();
    let mut infos: Vec<(String, String, String, Map<String, Value>)> = Vec::new();
    for (ws_id, entry) in all_config
        .iter()
        .filter(|(k, v)| k.as_str() != GLOBAL_CONFIG_KEY && v.is_object())
    {
        let display = crate::config::workspace_display_name(entry.get("name"), ws_id).to_string();
        let base = entry
            .get("worktree_base")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        let jobs = entry
            .get("jobs")
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default();
        jobs_by_name.insert(display.clone(), jobs.clone());
        infos.push((ws_id.clone(), display, base, jobs));
    }
    infos.sort_by(|a, b| a.0.cmp(&b.0));

    let mut result = Map::new();
    for (_ws_id, display, base, own_jobs) in infos {
        let ws_jobs = if base.is_empty() {
            own_jobs
        } else {
            // worktree はベースのジョブ定義を引き継ぐ（ベース未登録なら自前のまま）
            jobs_by_name.get(&base).cloned().unwrap_or(own_jobs)
        };
        result.insert(
            display,
            Value::Object(merge_jobs_serialized(&ws_jobs, &common_jobs)),
        );
    }
    // 動的 worktree（config 未登録）はベースのジョブ定義を引き継ぐ
    for wt in dynamic_worktree_entries(&state.config, None, false).await {
        let name = wt
            .get("name")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        if name.is_empty() || result.contains_key(&name) {
            continue;
        }
        let base = wt
            .get("worktree_base")
            .and_then(Value::as_str)
            .unwrap_or("");
        let wt_jobs = jobs_by_name.get(base).cloned().unwrap_or_default();
        result.insert(
            name,
            Value::Object(merge_jobs_serialized(&wt_jobs, &common_jobs)),
        );
    }
    Ok(Json(Value::Object(result)))
}

// ─── ワークスペースジョブ ───────────────────────────────────────────────────

pub async fn list_workspace_jobs(
    State(state): State<Arc<AppState>>,
    Path(name): Path<String>,
    _auth: RequireAuth,
) -> Result<Json<Value>, ApiError> {
    resolve_workspace_path(&state.config, &name).await?;
    Ok(Json(Value::Object(serialize_workspace_jobs(&state, &name))))
}

/// ws_jobs_context 相当: パス解決 + ジョブ共有オーナーの解決 + データロード。
async fn ws_jobs_data(
    state: &Arc<AppState>,
    name: &str,
) -> Result<(String, Map<String, Value>), ApiError> {
    resolve_workspace_path(&state.config, name).await?;
    let owner = resolve_jobs_owner(state, name);
    let data = load_workspace_jobs_data(state, &owner);
    Ok((owner, data))
}

pub async fn create_workspace_job(
    State(state): State<Arc<AppState>>,
    Path(name): Path<String>,
    _auth: RequireAuth,
    JsonBody(body): JsonBody<JobRequest>,
) -> Result<Json<Value>, ApiError> {
    let (owner, data) = ws_jobs_data(&state, &name).await?;
    let log_context = format!("job created workspace={name}");
    save_job(
        &state,
        data,
        move |s, m| commit_workspace_jobs(s, &owner, m),
        None,
        &body,
        &log_context,
    )
    .await
    .map(Json)
}

pub async fn reorder_workspace_jobs(
    State(state): State<Arc<AppState>>,
    Path(name): Path<String>,
    _auth: RequireAuth,
    JsonBody(body): JsonBody<ReorderJobsRequest>,
) -> Result<Json<Value>, ApiError> {
    let (owner, _data) = ws_jobs_data(&state, &name).await?;
    let log_context = format!("jobs reordered workspace={name}");
    reorder_jobs(
        &state,
        move |s, m| commit_workspace_jobs(s, &owner, m),
        &body.order,
        &log_context,
    )
    .map(Json)
}

pub async fn update_workspace_job(
    State(state): State<Arc<AppState>>,
    Path((name, job_name)): Path<(String, String)>,
    _auth: RequireAuth,
    JsonBody(body): JsonBody<JobRequest>,
) -> Result<Json<Value>, ApiError> {
    let (owner, data) = ws_jobs_data(&state, &name).await?;
    if !data.contains_key(&job_name) {
        return Err(not_found(format!("Job '{job_name}' not found")));
    }
    let log_context = format!("job updated workspace={name}");
    save_job(
        &state,
        data,
        move |s, m| commit_workspace_jobs(s, &owner, m),
        Some(job_name),
        &body,
        &log_context,
    )
    .await
    .map(Json)
}

pub async fn delete_workspace_job(
    State(state): State<Arc<AppState>>,
    Path((name, job_name)): Path<(String, String)>,
    _auth: RequireAuth,
) -> Result<Json<Value>, ApiError> {
    let (owner, _data) = ws_jobs_data(&state, &name).await?;
    let log_context = format!("job deleted workspace={name}");
    delete_job(
        &state,
        move |s, m| commit_workspace_jobs(s, &owner, m),
        &job_name,
        &format!("Job '{job_name}' not found"),
        &log_context,
    )
    .map(Json)
}

// ─── 共通ジョブ ─────────────────────────────────────────────────────────────

pub async fn list_common_jobs(
    State(state): State<Arc<AppState>>,
    _auth: RequireAuth,
) -> Result<Json<Value>, ApiError> {
    let data = load_common_jobs_data(&state);
    let mut out = Map::new();
    for (name, entry) in &data {
        out.insert(
            name.clone(),
            crate::jobs_common::job_entry_to_dict(name, entry, None),
        );
    }
    Ok(Json(Value::Object(out)))
}

pub async fn create_common_job(
    State(state): State<Arc<AppState>>,
    _auth: RequireAuth,
    JsonBody(body): JsonBody<JobRequest>,
) -> Result<Json<Value>, ApiError> {
    let data = load_common_jobs_data(&state);
    save_job(
        &state,
        data,
        commit_common_jobs,
        None,
        &body,
        "common job created",
    )
    .await
    .map(Json)
}

pub async fn update_common_job(
    State(state): State<Arc<AppState>>,
    Path(job_name): Path<String>,
    _auth: RequireAuth,
    JsonBody(body): JsonBody<JobRequest>,
) -> Result<Json<Value>, ApiError> {
    let data = load_common_jobs_data(&state);
    if !data.contains_key(&job_name) {
        return Err(not_found(format!("Common job '{job_name}' not found")));
    }
    save_job(
        &state,
        data,
        commit_common_jobs,
        Some(job_name),
        &body,
        "common job updated",
    )
    .await
    .map(Json)
}

pub async fn delete_common_job(
    State(state): State<Arc<AppState>>,
    Path(job_name): Path<String>,
    _auth: RequireAuth,
) -> Result<Json<Value>, ApiError> {
    delete_job(
        &state,
        commit_common_jobs,
        &job_name,
        &format!("Common job '{job_name}' not found"),
        "common job deleted",
    )
    .map(Json)
}

pub async fn reorder_common_jobs(
    State(state): State<Arc<AppState>>,
    _auth: RequireAuth,
    JsonBody(body): JsonBody<ReorderJobsRequest>,
) -> Result<Json<Value>, ApiError> {
    reorder_jobs(
        &state,
        commit_common_jobs,
        &body.order,
        "common jobs reordered",
    )
    .map(Json)
}
